src/server/game/Entities/Unit/Unit.h

```diff
@@ -339,6 +311,30 @@ enum TriggerCastFlags : uint32
     TRIGGERED_FULL_DEBUG_MASK                       = 0xFFFFFFFF
 };
 
+struct CastSpellExtraArgs
+{
+    CastSpellExtraArgs() {}
+    CastSpellExtraArgs(bool triggered) : triggerFlags(triggered ? TRIGGERED_FULL_MASK : TRIGGERED_NONE) {}
+    CastSpellExtraArgs(TriggerCastFlags trigger) : triggerFlags(trigger) {}
+    CastSpellExtraArgs(AuraEffect const* eff) : triggerFlags(TRIGGERED_FULL_MASK), triggeredByAura(eff) {}
+    CastSpellExtraArgs(ObjectGuid const& origCaster) : triggerFlags(TRIGGERED_FULL_MASK), originalCaster(origCaster) {}
+
+    TriggerCastFlags triggerFlags = TRIGGERED_NONE;
+    Item* castItem = nullptr;
+    AuraEffect const* triggeredByAura = nullptr;
+    ObjectGuid originalCaster = ObjectGuid::Empty;
+    struct
+    {
+        void add(SpellValueMod mod, int32 val) { data.push_back({ mod, val }); }
+        void add_bp0(int32 bp0) { add(SPELLVALUE_BASE_POINT0, bp0); } // because i don't want to type SPELLVALUE_BASE_POINT0 300 times
+
+        auto begin() const { return data.cbegin(); }
+        auto end() const { return data.cend(); }
+
+        std::vector<std::pair<SpellValueMod, int32>> data;
+    } customValues;
```

Comment:
can we change this to allow method chaining?
```c++
CastSpellExtraArgs()
    .setTriggered()
    .addSpellValue(SPELLVALUE_BASE_POINT0, value)
    .setOriginalCaster(caster)
```

Reply:
meh, i'm not particularly fond of it. Starts to get unreadable very quickly, and there's little benefit over the existing syntax.

Reply:
Why not?
But both of you, dont forget TC codestyle - functions and public struct/class data members are start with uppercase

Reply:
> starts to get unreadable very quickly
The primary use case is putting something like @joschiwald posted directly in the invocation. This makes for some very unreadable code.

Reply:
first time i saw that method chaining syntax, it was also a bit bad readable to me, but now i like it

to support chaining does not mean you must use it, however you like

Reply:
it's not a question of whether _i_ would use it, it's a question of whether we want code using it in the core.

Reply:
Look at MailDraft uses, I think it works pretty well